buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginBuildPlugin.java

```diff
@@ -0,0 +1,332 @@
+        dependencies.add("compileOnly", "org.locationtech.spatial4j:spatial4j:" + VersionProperties.getVersions().get("spatial4j"));
+        dependencies.add("compileOnly", "org.locationtech.jts:jts-core:" + VersionProperties.getVersions().get("jts"));
+        dependencies.add("compileOnly", "org.apache.logging.log4j:log4j-api:" + VersionProperties.getVersions().get("log4j"));
+        dependencies.add("compileOnly", "org.apache.logging.log4j:log4j-core:" + VersionProperties.getVersions().get("log4j"));
+        dependencies.add("compileOnly", "org.elasticsearch:jna:" + VersionProperties.getVersions().get("jna"));
+    }
+
+    /**
+     * Adds a bundlePlugin task which builds the zip containing the plugin jars,
+     * metadata, properties, and packaging files
+     */
+    private static TaskProvider<Zip> createBundleTasks(final Project project, PluginPropertiesExtension extension) {
+        final var pluginMetadata = project.file("src/main/plugin-metadata");
+        final var templateFile = new File(project.getBuildDir(), "templates/plugin-descriptor.properties");
+
+        // create tasks to build the properties file for this plugin
+        final var copyPluginPropertiesTemplate = project.getTasks().register("copyPluginPropertiesTemplate", new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                task.getOutputs().file(templateFile);
+                // intentionally an Action and not a lambda to avoid issues with up-to-date check
+                task.doLast(new Action<Task>() {
+                    @Override
+                    public void execute(Task task) {
+                        InputStream resourceTemplate = PluginBuildPlugin.class.getResourceAsStream("/" + templateFile.getName());
+                        try {
+                            String templateText = IOUtils.toString(resourceTemplate, StandardCharsets.UTF_8.name());
+                            FileUtils.write(templateFile, templateText, "UTF-8");
+                        } catch (IOException e) {
+                            throw new GradleException("Unable to copy plugin properties", e);
+                        }
+                    }
+                });
+            }
+        });
+        final var buildProperties = project.getTasks().register("pluginProperties", Copy.class, copy -> {
+            copy.dependsOn(copyPluginPropertiesTemplate);
+            copy.from(templateFile);
+            copy.into(new File(project.getBuildDir(), "generated-resources"));
+        });
+        // add the plugin properties and metadata to test resources, so unit tests can
+        // know about the plugin (used by test security code to statically initialize the plugin in unit tests)
+        var testSourceSet = project.getExtensions().getByType(SourceSetContainer.class).getByName("test");
+        Map<String, Object> map = new LinkedHashMap<String, Object>(1);
+        map.put("builtBy", buildProperties);
```

Comment:
Can we just do this, or does the map need to be mutable?
```suggestion
        Map<String, Object> map = Map.of("builtBy", buildProperties);
```